# E-Commerce Logs Analysis

## Project Objective

The objective of this project is to analyze the website logs from an e-commerce store to obtain useful insight into customer behaviour.

## Data Source

The dataset was obtained here: https://www.kaggle.com/datasets/kzmontage/e-commerce-website-logs and the data has an Open Database license: https://opendatacommons.org/licenses/odbl/1-0/

In [26]:
-- Specifying database
USE EcommerceLogsDb

Commands completed successfully.

Total execution time: 00:00:00.003

In [27]:
-- preview of purchase logs table
SELECT TOP 3*
FROM logs;

(3 rows affected)

Total execution time: 00:00:00.008

access_date duration_seconds Proto IP Src_IP_type Src_Pt Bytes Accessed_From country membership Languages Sales Returned Returned_Amount 2016-11-01 09:58:00.0000000 2533 TCP 1.10.195.126 EXT_SERVER 8082 20100 Chrome Albania Normal Albanian 261.96 No 0 2016-11-01 09:59:00.0000000 4034 TCP 1.1.217.211 OPENSTACK_NET 56978 20500 Mozilla Firefox Albania Normal Albanian 731.94 No 0 2016-11-01 09:59:00.0000000 1525 TCP 1.115.198.107 EXT_SERVER 8082 90100 Mozilla Firefox Albania Normal Albanian 14.62 No 0

In [28]:
-- Which country had the highest number of sales and highest value of sales (in $)?
-- showing country percentiles for both metrics using window functions
SELECT country, 
    COUNT(*) AS num_sales, 
    PERCENT_RANK() OVER(ORDER BY COUNT(*)) AS num_sales_percentile, 
    SUM(Sales) AS value_sales, 
    PERCENT_RANK() OVER(ORDER BY SUM(Sales)) AS value_sales_percentile
FROM logs
GROUP BY country
ORDER BY value_sales DESC;

(101 rows affected)

Total execution time: 00:00:00.139

country num_sales num_sales_percentile value_sales value_sales_percentile Austria 2674 1 1296463.3414999906 1 Belgium 2604 0.97 1154321.3705999977 0.99 Argentina 2603 0.96 1140238.6927999987 0.98 Australia 2520 0.94 1139576.1649999968 0.97 United States 2231 0.73 1132496.3800000001 0.96 Brazil 2604 0.97 1129919.2000999975 0.95 Norway 2160 0.68 1129793.889999999 0.94 Netherlands 2292 0.88 1106623.810000002 0.93 Spain 2232 0.74 1102578.460000004 0.92 Puerto Rico 2232 0.74 1089656.3800000006 0.91 Canada 2436 0.92 1078310.4730000005 0.9 United Kingdom 2232 0.74 1069649.1000000013 0.89 Bulgaria 2520 0.94 1068255.9588000018 0.88 Republic of Korea 2232 0.74 1063598.8800000008 0.87 Singapore 2232 0.74 1062442.5199999993 0.86 Suriname 2016 0.58 1037655.410000003 0.85 Mexico 2232 0.74 1034845.1700000012 0.84 Turkmenistan 2088 0.63 1032696.5800000039 0.83 Belize 2352 0.91 1009231.1078000019 0.82 Sweden 2148 0.67 1006956.3000000026 0.81 Chile 2604 0.97 998301.8032000033 0.8 Russia 1952 0.55 990272.1340000018 0.79 Thailand 2004 0.56 984841.0600000018 0.78 Portugal 1944 0.52 955279.5800000016 0.77 Saint Lucia 2016 0.58 953605.8500000038 0.76 Mauritius 2292 0.88 951543.6300000029 0.75 Trinidad and Tobago 1944 0.52 940220.2900000017 0.74 Antigua and Barbuda 2268 0.87 937031.6729000005 0.73 Romania 2004 0.56 933676.0400000019 0.72 Armenia 2086 0.62 918236.707500002 0.71 Barbados 2099 0.66 914011.7824000027 0.7 Uruguay 2016 0.58 898636.5600000003 0.69 New Zealand 2088 0.63 897640.3300000007 0.68 Paraguay 1944 0.52 887757.4500000015 0.67 Luxembourg 2232 0.74 869798.3930000014 0.66 Ukraine 2016 0.58 860557.6200000015 0.65 Bahamas 1932 0.5 858137.5198000026 0.64 Malta 2232 0.74 849198.6836000021 0.63 Panama 1800 0.41 827407.4100000006 0.62 Poland 1728 0.35 823874.8300000004 0.61 Colombia 2514 0.93 823458.3767000027 0.6 Albania 1848 0.44 816714.1692000002 0.59 Guatemala 2160 0.68 811037.0116000031 0.58 Bahrain 1764 0.39 800315.182500001 0.57 Belarus 1763 0.38 793122.7560000019 0.56 Saint Vincent and Grenadines 1800 0.41 782619.9400000017 0.55 Greece 2232 0.74 781112.8694000007 0.54 Slovenia 1511 0.25 779910.2300000001 0.53 Slovakia 1584 0.31 774529.1199999993 0.52 Switzerland 1512 0.27 711132.8300000015 0.51 Ecuador 2232 0.74 704434.9145000005 0.5 Uzbekistan 1584 0.31 700289.7400000006 0.49 South Africa 1440 0.24 699497.384000001 0.48 Japan 2233 0.86 692880.2690000008 0.47 Israel 2232 0.74 678394.283000001 0.46 Grenada 1860 0.45 677864.1923000009 0.45 Iceland 2292 0.88 673693.6802000001 0.44 Guyana 1799 0.4 673308.8582000005 0.43 Kuwait 1800 0.41 668267.5795000002 0.42 Kyrgyzstan 1872 0.47 666682.2039000005 0.41 Ireland 2160 0.68 649537.6800000011 0.4 France 2160 0.68 643784.2059999998 0.39 Kazakhstan 1872 0.47 641065.9040000006 0.38 Hungary 1860 0.45 634635.7702 0.37 Costa Rica 2160 0.68 632901.6622000004 0.36 Finland 2088 0.63 630622.2634000012 0.35 Italy 2232 0.74 625664.3300000008 0.34 Seychelles 1296 0.21 620275.4700000003 0.33 Azerbaijan 1344 0.23 610523.8879999998 0.32 Germany 1872 0.47 601330.1426000008 0.31 El Salvador 1728 0.35 593740.2684 0.3 Lithuania 1572 0.29 572731.2999999988 0.29 Serbia 1296 0.21 564452.1499999986 0.28 Qatar 1068 0.17 554514.8799999999 0.27 Latvia 1511 0.25 549710.9939999995 0.26 Aruba 1176 0.19 537197.9709999991 0.25 Philippines 1079 0.18 522248.86000000004 0.24 Czech Republic 1932 0.5 517375.4399999993 0.23 Cuba 1728 0.35 500943.9629999991 0.22 Georgia 1584 0.31 499963.55559999857 0.21 Denmark 1584 0.31 449218.213999999 0.2 Estonia 1512 0.27 424996.7387999989 0.19 Croatia 1572 0.29 416275.25939999963 0.18 Sri Lanka 792 0.13 395918.2999999992 0.17 Montenegro 720 0.11 357511.14999999985 0.16 Jamaica 1224 0.2 347578.4049999999 0.15 Kiribati 792 0.13 323550.16689999966 0.14 Cyprus 1067 0.16 281314.48579999973 0.13 Turkey 504 0.1 280994.1200000001 0.12 Nicaragua 432 0.08 223247.22000000006 0.11 Maldives 720 0.11 217764.10799999998 0.1 United Arab Emirates 432 0.08 186636.71000000022 0.09 Fiji 792 0.13 186410.4449999999

From the table above, the top 4 countries for num\_sales and value\_sales are Austria, Belgium, Argentina and Australia. The number of sales seems highly correlated with value of sales with some exceptions like the United States which ranks in the 73rd percentile for number of sales and in the 96th percentile for value of sales.

Next let's look at the average order value (AOV) per country and compare to the worldwide average. The AOV is a popular e-commerce metric that indicates the average amount customers are spending in one purchase. Both the tradditional AOV (using the mean) and the median AOV will be calculated adnc compared.

In [29]:
-- average order value ($) per country
SELECT  
    country, AVG(Sales) AS avg_order_value,
    -- subquery in SELECT to determine worldwide avg order value ($)
    (SELECT AVG(Sales) FROM logs) AS world_avg_order_value,
    -- difference bewteen country avg and worldwide avg with another SELECT subquery
    (AVG(Sales) - (SELECT AVG(Sales) FROM logs)) AS difference
FROM logs
GROUP BY country
ORDER BY difference DESC;

(101 rows affected)

Total execution time: 00:00:00.108

country avg_order_value world_avg_order_value difference Turkey 557.5280158730159 411.3464487919873 146.1815670810314 Norway 523.0527268518516 411.3464487919873 111.70627805986715 Qatar 519.2086891385769 411.3464487919873 107.86224034659244 Nicaragua 516.7759722222223 411.3464487919873 105.42952343023785 Slovenia 516.1550165453345 411.3464487919873 104.80856775335008 Suriname 514.710024801587 411.3464487919873 103.36357600960253 United States 507.6182787987447 411.3464487919873 96.27183000676024 Russia 507.31154405737686 411.3464487919873 95.96509526539239 Sri Lanka 499.89684343434334 411.3464487919873 88.55039464235887 Montenegro 496.54326388888893 411.3464487919873 85.19681509690446 Turkmenistan 494.5864846743293 411.3464487919873 83.24003588234484 Spain 493.9867652329749 411.3464487919873 82.64031644099043 Thailand 491.4376546906187 411.3464487919873 80.0912058986342 Portugal 491.3989609053496 411.3464487919873 80.05251211336514 Slovakia 488.9704040404043 411.3464487919873 77.6239552484198 Oman 488.639675925926 411.3464487919873 77.29322713394151 Puerto Rico 488.1973028673829 411.3464487919873 76.85085407539844 South Africa 485.76207222222246 411.3464487919873 74.41562343023799 Austria 484.84044184741964 411.3464487919873 73.49399305543517 Philippines 484.011918443003 411.3464487919873 72.66546965101855 Trinidad and Tobago 483.6524125514405 411.3464487919873 72.305963759456 Netherlands 482.8201614310641 411.3464487919873 71.47371263907962 United Kingdom 479.23346774193544 411.3464487919873 67.88701894995097 Seychelles 478.6076157407405 411.3464487919873 67.26116694875606 Poland 476.77941550925937 411.3464487919873 65.4329667172749 Republic of Korea 476.52279569892426 411.3464487919873 65.17634690693978 Singapore 476.00471326164853 411.3464487919873 64.65826446966406 Saint Lucia 473.01877480158737 411.3464487919873 61.6723260096029 Switzerland 470.32594576719595 411.3464487919873 58.979496975211475 Sweden 468.78784916201096 411.3464487919873 57.441400370026486 Romania 465.9062075848299 411.3464487919873 54.5597587928454 Mexico 463.6403091397846 411.3464487919873 52.29386034780015 Panama 459.6707833333333 411.3464487919873 48.32433454134883 Aruba 456.80099574829944 411.3464487919873 45.45454695631497 Paraguay 456.6653549382713 411.3464487919873 45.31890614628685 Azerbaijan 454.2588452380952 411.3464487919873 42.91239644611073 San Marino 453.7129166666668 411.3464487919873 42.366467874682314 Bahrain 453.69341411564625 411.3464487919873 42.34696532366178 Australia 452.2127638888888 411.3464487919873 40.866315096904316 Belarus 449.87110380034005 411.3464487919873 38.52465500835558 Uruguay 445.7522619047622 411.3464487919873 34.40581311277771 Bahamas 444.1705589026916 411.3464487919873 32.824110110707124 Belgium 443.28777672811043 411.3464487919873 31.94132793612596 Canada 442.65618760262726 411.3464487919873 31.309738810642784 Uzbekistan 442.10210858585845 411.3464487919873 30.755659793873974 Albania 441.94489675324706 411.3464487919873 30.598447961262593 Armenia 440.1901761744965 411.3464487919873 28.843727382512043 Argentina 438.04790349596607 411.3464487919873 26.701454703981597 Serbia 435.53406635802514 411.3464487919873 24.18761756604067 Barbados 435.45106355407324 411.3464487919873 24.104614762088772 Cabo Verde 435.26276190476193 411.3464487919873 23.91631311277746 Saint Vincent and Grenadines 434.78885555555576 411.3464487919873 23.442406763571285 Brazil 433.91674350998454 411.3464487919873 22.57029471800007 United Arab Emirates 432.02942129629633 411.3464487919873 20.682972504311863 New Zealand 429.90437260536385 411.3464487919873 18.557923813379375 Belize 429.09485875850305 411.3464487919873 17.74840996651858 Ukraine 426.8638988095237 411.3464487919873 15.517450017539204 Bulgaria 423.9110947619046 411.3464487919873 12.56464596992015 Mauritius 415.1586518324604 411.3464487919873 3.8122030404759357 Antigua and Barbuda 413.15329492945375 411.3464487919873 1.8068461374692788 Kiribati 408.52293800505043 411.3464487919873 -2.

In [45]:
-- median order value ($) per country
WITH CTE_AOV AS (
    SELECT  
        DISTINCT country,
        PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY Sales) OVER(PARTITION BY country) AS median_AOV,
        PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY Sales) OVER() AS world_median_AOV
    FROM logs)
    --ORDER BY median_AOV DESC)

SELECT DISTINCT logs.country, median_AOV, AVG(median_AOV) OVER () AS avg_world_aov
FROM logs
    INNER JOIN CTE_AOV
    ON logs.country = CTE_AOV.country;

(101 rows affected)

Total execution time: 00:00:01.842

country median_AOV avg_world_aov Sweden 75.75 57.02802231569449 Puerto Rico 75.75 57.02802231569449 Finland 26.15 57.02802231569449 Japan 15.552 57.02802231569449 Brazil 71.68 57.02802231569449 Guyana 40.66 57.02802231569449 Switzerland 75.75 57.02802231569449 San Marino 121.98 57.02802231569449 Austria 81.32 57.02802231569449 Maldives 17.808 57.02802231569449 Thailand 71.68 57.02802231569449 Mexico 75.75 57.02802231569449 South Africa 81.32 57.02802231569449 Saint Kitts and Nevis 71.68 57.02802231569449 Italy 15.69 57.02802231569449 Lithuania 35.84 57.02802231569449 Bulgaria 71.68 57.02802231569449 Serbia 81.32 57.02802231569449 Belize 71.68 57.02802231569449 Fiji 14.075 57.02802231569449 Kyrgyzstan 35.599999999999994 57.02802231569449 Australia 71.68 57.02802231569449 Cabo Verde 60.6 57.02802231569449 Oman 58.65 57.02802231569449 Netherlands 71.68 57.02802231569449 Luxembourg 40.66 57.02802231569449 Trinidad and Tobago 81.32 57.02802231569449 Spain 81.32 57.02802231569449 Latvia 40.66 57.02802231569449 Dominica 10.46 57.02802231569449 Bahrain 75.75 57.02802231569449 Ukraine 71.68 57.02802231569449 Armenia 71.68 57.02802231569449 Norway 81.32 57.02802231569449 Paraguay 60.6 57.02802231569449 Belgium 73.875 57.02802231569449 Saint Vincent and Grenadines 66.14 57.02802231569449 Aruba 81.32 57.02802231569449 Philippines 58.65 57.02802231569449 Mauritius 58.65 57.02802231569449 Croatia 23.46 57.02802231569449 Qatar 81.32 57.02802231569449 Azerbaijan 75.75 57.02802231569449 Costa Rica 25.875 57.02802231569449 Antigua and Barbuda 60.6 57.02802231569449 Cuba 29.04 57.02802231569449 Republic of Korea 75.75 57.02802231569449 Poland 60.6 57.02802231569449 Suriname 81.32 57.02802231569449 Israel 23.39 57.02802231569449 Slovenia 81.32 57.02802231569449 Slovakia 81.32 57.02802231569449 Singapore 75.75 57.02802231569449 Ecuador 35.19 57.02802231569449 Georgia 40.56 57.02802231569449 Russia 75.75 57.02802231569449 Denmark 21.784 57.02802231569449 Kiribati 46.92 57.02802231569449 Hungary 26.15 57.02802231569449 New Zealand 58.65 57.02802231569449 Romania 81.32 57.02802231569449 United States 81.32 57.02802231569449 Panama 60.6 57.02802231569449 Portugal 81.32 57.02802231569449 Turkey 75.75 57.02802231569449 Argentina 71.68 57.02802231569449 Canada 71.68 57.02802231569449 Nicaragua 81.32 57.02802231569449 Guatemala 40.66 57.02802231569449 Macau 35.515 57.02802231569449 Cyprus 21.93 57.02802231569449 Colombia 29.3115 57.02802231569449 Seychelles 75.75 57.02802231569449 Montenegro 81.32 57.02802231569449 Uzbekistan 71.68 57.02802231569449 Belarus 71.68 57.02802231569449 Barbados 75.75 57.02802231569449 Uruguay 71.68 57.02802231569449 France 28.831 57.02802231569449 Albania 71.68 57.02802231569449 Greece 30.3 57.02802231569449 United Arab Emirates 66.14 57.02802231569449 Germany 41.352000000000004 57.02802231569449 Grenada 36.343999999999994 57.02802231569449 Malta 40.66 57.02802231569449 Iceland 17.019 57.02802231569449 Kuwait 40.66 57.02802231569449 El Salvador 40.66 57.02802231569449 Jamaica 13.194 57.02802231569449 Estonia 17.92 57.02802231569449 Bosnia and Herzegovina 78.535 57.02802231569449 Czech Republic 24.064999999999998 57.02802231569449 United Kingdom 71.68 57.02802231569449 Saint Lucia 71.68 57.02802231569449 Mongolia 94.41999999999999 57.02802231569449 Chile 41.91 57.02802231569449 Bahamas 75.75 57.02802231569449 Ireland 24.695 57.02802231569449 Sri Lanka 75.75 57.02802231569449 Turkmenistan 75.75 57.02802231569449 Kazakhstan 31.14 57.02802231569449

The average order values (mean) range from $235-$557 and are much higher than the median AOV's which range from $10-$122. This indicates that there are several large outliers for each dataset that pull the mean much higher than the median. As a result, we will use the median AOV because it is more robust to outliers and is a better representation for central tendency in this case.

We'll now look at returns of purchased items for each country.

In [35]:
-- CTE for value of sales returned
WITH returned_value AS (
    SELECT country,
        -- Sum of total dollars returned
        SUM(Returned_Amount) AS returned_amount,
        -- Percentage of dollars returned over dollars sold
        -- Sales column doesn't include amount for returned items
        SUM(Returned_Amount) / (SUM(Sales) + SUM(Returned_Amount)) AS pct_value_returned
    From logs
    GROUP BY country),

-- CTE for number of sales returned
returned_sales AS (
    SELECT t1.country, 
        -- counting all rows (number of transactions)
        COUNT(*) AS num_sales,
        -- counting number of order returns 
        COUNT(t2.Returned) AS sales_returned,
        -- calculating percentage of orders that were returned
        1.0 * COUNT(t2.Returned) / COUNT(*) AS pct_sales_returned
    FROM logs as t1
    LEFT JOIN (SELECT duration_seconds, ip, country, Returned
    FROM logs 
    WHERE Returned = 'Yes') AS t2
    ON t1.duration_seconds = t2.duration_seconds
    AND t1.ip = t2.ip
    GROUP BY t1.country)

-- final query with percent of orders returned and percent of sales refunded
SELECT CTE1.country, pct_sales_returned, pct_value_returned 
FROM returned_value AS CTE1
    INNER JOIN returned_sales AS CTE2
    ON CTE1.country = CTE2.country 
ORDER BY pct_sales_returned DESC, pct_value_returned DESC;



Warning: Null value is eliminated by an aggregate or other SET operation.

(101 rows affected)

Total execution time: 00:00:00.764

country pct_sales_returned pct_value_returned Nicaragua 0.212962962962 0.21179128180131432 Oman 0.212962962962 0.15195254761361243 Montenegro 0.190277777777 0.21973417846937762 Puerto Rico 0.184139784946 0.15605512400495494 Poland 0.182870370370 0.1416946334543681 Paraguay 0.178497942386 0.17183279571242158 Philippines 0.177015755329 0.15597966717691733 New Zealand 0.171455938697 0.16719107551492732 Qatar 0.171348314606 0.13725544107967674 Portugal 0.170781893004 0.15266371014340657 Republic of Korea 0.170698924731 0.1386451845251341 Turkmenistan 0.170498084291 0.1755562065563355 Norway 0.168055555555 0.14648826008177862 Ukraine 0.161706349206 0.15800120178596974 Slovakia 0.157828282828 0.13235868024179404 Panama 0.157777777777 0.161759685174265 Saint Kitts and Nevis 0.157407407407 0.11034302063778338 Switzerland 0.156746031746 0.1838030632666294 Serbia 0.156635802469 0.18601890985154315 Dominica 0.152777777777 0.41682638534275407 South Africa 0.152777777777 0.11065572592410283 Saint Lucia 0.152281746031 0.15103533388885867 Netherlands 0.152268760907 0.15423967382326148 Mexico 0.151881720430 0.1585345599836336 Mauritius 0.150959860383 0.16927232672302803 United Arab Emirates 0.150462962962 0.2207978693272318 Mongolia 0.150000000000 0.21720911162519385 Thailand 0.149700598802 0.14140295636411362 Trinidad and Tobago 0.149691358024 0.166058966984402 Uruguay 0.149305555555 0.15546877980131008 Seychelles 0.148919753086 0.15288077020941138 Saint Vincent and Grenadines 0.147222222222 0.1578306134398837 United Kingdom 0.146887595163 0.16909418208500715 Sweden 0.146648044692 0.159694984331105 Albania 0.144480519480 0.17508950794223882 San Marino 0.143518518518 0.08445700058144673 Singapore 0.143369175627 0.14644639878451488 Latvia 0.142951687624 0.17880813190249115 Suriname 0.142857142857 0.12274140160575248 United States 0.141640519946 0.13763357599841128 Guyana 0.140633685380 0.20442932048887963 Armenia 0.139980824544 0.16414319749023445 Spain 0.138440860215 0.11356458673045527 Romania 0.138223552894 0.13853069420367586 Belize 0.136479591836 0.16056040871036592 Bahrain 0.136054421768 0.13987344599350557 Kuwait 0.136035535813 0.1525867712832713 Belarus 0.135564378899 0.14206112724429593 Slovenia 0.135009927200 0.09834957001717466 Barbados 0.134826107670 0.13714564874899382 Brazil 0.134792626728 0.16863282966199297 Bahamas 0.134505949301 0.1442156490858075 Argentina 0.133307721859 0.1619016520445564 Uzbekistan 0.132575757575 0.16599578790261404 Guatemala 0.131944444444 0.17122064074310195 Russia 0.131147540983 0.1410352114251003 Turkey 0.130952380952 0.10766988675118998 Luxembourg 0.130824372759 0.15466121778979863 Antigua and Barbuda 0.130511463844 0.1671689258731355 Aruba 0.129251700680 0.14296296034653322 Australia 0.127777777777 0.1342574598171063 Sri Lanka 0.127525252525 0.09093712300080514 Hungary 0.126881720430 0.20215967647807814 Azerbaijan 0.126488095238 0.1253355526735021 Maldives 0.126388888888 0.2182148399008438 Belgium 0.124807987711 0.15217351915917593 Greece 0.124551971326 0.1855260940375158 Austria 0.124158563949 0.13609233836509788 Kyrgyzstan 0.119123931623 0.14385089069668963 Grenada 0.117741935483 0.1613112295238262 Lithuania 0.117684478371 0.12114375200106973 Malta 0.116039426523 0.14876467468035276 Kazakhstan 0.113247863247 0.15420071849191674 France 0.112500000000 0.17091972743941955 Bulgaria 0.111507936507 0.11576664685882165 Kiribati 0.111111111111 0.11578842082363633 Italy 0.107974910394 0.1778600216336717 Chile 0.107910906298 0.15475739181808754 Iceland 0.106893542757 0.1806826406781312 Ireland 0.104629629629 0.18280738764487423 Canada 0.104269293924 0.1387213302940069 Israel 0.102598566308 0.16731406262902562 Japan 0.102552619793 0.13061369704727427 Jamaica 0.102124183006 0.12022248900311977 Ecuador 0.101702508960 0.16961774067380567 Fiji 0.101010101010 0.2251059913195575 Germany 0.100961538461 0.17240300683900145 Georgia 0.099116161616 0.16199832378698306 El Salvador 0.097800925925 0.14052520722058312 Finlan

Nicaragua and Oman have the highest percentage of orders returned at 21.3%. This represents a possible opportunity to increase revenue by making the return policy more strict to attempt to lower the number of returned orders. The affect that a stricter return policy would have on customer experience should also be taken into account though.

We will next look at the number of returns for each language. This will help inform us of any language/communication issues thay may be causing a higher rate of purchase returns.

In [32]:
SELECT TOP 5 t1.Languages, 
        -- calculating percentage of orders that were returned
        ROUND((1.0 * COUNT(t2.Returned) / COUNT(*)), 2) AS pct_orders_returned
FROM logs as t1
        LEFT JOIN (SELECT duration_seconds, ip, country, Returned
        FROM logs 
        WHERE Returned = 'Yes') AS t2
        ON t1.duration_seconds = t2.duration_seconds
        AND t1.ip = t2.ip
GROUP BY t1.Languages
ORDER BY pct_orders_returned DESC;

Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.102

Languages pct_orders_returned marathi 0.220000000000 romanian 0.180000000000 kinyarwanda 0.180000000000 polish 0.170000000000 Slovak 0.170000000000

Marathi, a language spoken in India, has the highest percentage for orders returned (4% above the second language). A statistical test should be performed to determine if this percentage is statistically significantly greater than the 2nd highest percentage. If it is, this indicates that higher return rates could be linked to translation issues with the language. This means customers could be misinformed about products, resulting in higher return rates. The impact of language on return rates should be further investigated for the website as a whole to idenitfy if that is a frequent issue.

The membership types associated with online purchases will now be investigated per country. It is important to note that each count of a membership type represents an online purchase but not a unique individual (no unique identifiers are present in dataset). It is likely that premium members make more frequent purchases than normal and non-members. This category therefore investigates purchases per membership type and not individuals per membership type.

In [33]:
-- first CTE for total website purchases per country
WITH totals AS (
    SELECT country, COUNT(*) AS total_purchases
    FROM logs
    GROUP BY country),

-- second CTE for website purchases per membership type per country
memberships AS (
    SELECT country, membership, COUNT(1) AS website_purchase
    FROM logs
    GROUP BY country, membership)

-- joining CTEs and calculating pct membership type for total purchases per country
SELECT memberships.country, membership, website_purchase, 1.0 * website_purchase/total_purchases AS pct_visits
FROM memberships
    LEFT JOIN totals
    ON memberships.country = totals.country
ORDER BY country;

(298 rows affected)

Total execution time: 00:00:00.084

country membership website_purchase pct_visits Albania Not Logged In 17 0.009199134199 Albania Premium 1185 0.641233766233 Albania Normal 646 0.349567099567 Antigua and Barbuda Normal 721 0.317901234567 Antigua and Barbuda Premium 1527 0.673280423280 Antigua and Barbuda Not Logged In 20 0.008818342151 Argentina Normal 967 0.371494429504 Argentina Premium 1612 0.619285439877 Argentina Not Logged In 24 0.009220130618 Armenia Not Logged In 20 0.009587727708 Armenia Normal 678 0.325023969319 Armenia Premium 1388 0.665388302972 Aruba Premium 763 0.648809523809 Aruba Not Logged In 4 0.003401360544 Aruba Normal 409 0.347789115646 Australia Normal 795 0.315476190476 Australia Premium 1684 0.668253968253 Australia Not Logged In 41 0.016269841269 Austria Premium 1827 0.683246073298 Austria Not Logged In 12 0.004487658937 Austria Normal 835 0.312266267763 Azerbaijan Not Logged In 6 0.004464285714 Azerbaijan Premium 914 0.680059523809 Azerbaijan Normal 424 0.315476190476 Bahamas Not Logged In 10 0.005175983436 Bahamas Normal 568 0.293995859213 Bahamas Premium 1354 0.700828157349 Bahrain Premium 1258 0.713151927437 Bahrain Not Logged In 11 0.006235827664 Bahrain Normal 495 0.280612244897 Barbados Premium 1402 0.667937112910 Barbados Normal 687 0.327298713673 Barbados Not Logged In 10 0.004764173415 Belarus Not Logged In 14 0.007941009642 Belarus Normal 546 0.309699376063 Belarus Premium 1203 0.682359614293 Belgium Normal 816 0.313364055299 Belgium Premium 1758 0.675115207373 Belgium Not Logged In 30 0.011520737327 Belize Normal 814 0.346088435374 Belize Not Logged In 6 0.002551020408 Belize Premium 1532 0.651360544217 Bosnia and Herzegovina Not Logged In 2 0.011904761904 Bosnia and Herzegovina Premium 122 0.726190476190 Bosnia and Herzegovina Normal 44 0.261904761904 Brazil Normal 931 0.357526881720 Brazil Not Logged In 16 0.006144393241 Brazil Premium 1657 0.636328725038 Bulgaria Not Logged In 34 0.013492063492 Bulgaria Normal 914 0.362698412698 Bulgaria Premium 1572 0.623809523809 Cabo Verde Normal 27 0.321428571428 Cabo Verde Premium 57 0.678571428571 Canada Normal 854 0.350574712643 Canada Not Logged In 28 0.011494252873 Canada Premium 1554 0.637931034482 Chile Normal 857 0.329109062980 Chile Not Logged In 249 0.095622119815 Chile Premium 1498 0.575268817204 Colombia Not Logged In 579 0.230310262529 Colombia Normal 848 0.337311058074 Colombia Premium 1087 0.432378679395 Costa Rica Not Logged In 522 0.241666666666 Costa Rica Premium 947 0.438425925925 Costa Rica Normal 691 0.319907407407 Croatia Normal 429 0.272900763358 Croatia Premium 765 0.486641221374 Croatia Not Logged In 378 0.240458015267 Cuba Not Logged In 350 0.202546296296 Cuba Premium 926 0.535879629629 Cuba Normal 452 0.261574074074 Cyprus Premium 517 0.484536082474 Cyprus Normal 292 0.273664479850 Cyprus Not Logged In 258 0.241799437675 Czech Republic Normal 509 0.263457556935 Czech Republic Not Logged In 513 0.265527950310 Czech Republic Premium 910 0.471014492753 Denmark Normal 370 0.233585858585 Denmark Not Logged In 390 0.246212121212 Denmark Premium 824 0.520202020202 Dominica Premium 26 0.361111111111 Dominica Not Logged In 21 0.291666666666 Dominica Normal 25 0.347222222222 Ecuador Premium 1152 0.516129032258 Ecuador Not Logged In 390 0.174731182795 Ecuador Normal 690 0.309139784946 El Salvador Not Logged In 299 0.173032407407 El Salvador Premium 870 0.503472222222 El Salvador Normal 559 0.323495370370 Estonia Normal 358 0.236772486772 Estonia Premium 683 0.451719576719 Estonia Not Logged In 471 0.311507936507 Fiji Normal 212 0.267676767676 Fiji Premium 316 0.398989898989 Fiji Not Logged In 264 0.333333333333 Finland Not Logged In 507 0.242816091954 Finland Normal 550 0.263409961685 Finland Premium 1031 0.493773946360 France Not Logged In 406 0.187962962962 France Normal 669 0.309722222222 France Premium 1085 0.502314814814 Georgia Not Logged In 223 0.140782828282 Georgia Premium 901 0.568813131313 Georgia Normal 460 0.290404040404 Germany Premium 982 0.524572649572 

In [34]:
-- purchases per browser per country
SELECT country, Accessed_From, COUNT(1) AS purchases
FROM logs
GROUP BY country, Accessed_From
ORDER BY country;

(695 rows affected)

Total execution time: 00:00:00.123

country Accessed_From purchases Albania Microsoft Edge 171 Albania Android App 392 Albania Chrome 299 Albania Safari 180 Albania IOS App 233 Albania Others 281 Albania Mozilla Firefox 292 Antigua and Barbuda IOS App 289 Antigua and Barbuda Microsoft Edge 189 Antigua and Barbuda Others 336 Antigua and Barbuda Chrome 401 Antigua and Barbuda Safari 210 Antigua and Barbuda Android App 486 Antigua and Barbuda Mozilla Firefox 357 Argentina Android App 575 Argentina IOS App 336 Argentina Chrome 429 Argentina Others 407 Argentina Safari 240 Argentina Microsoft Edge 216 Argentina Mozilla Firefox 400 Armenia Others 314 Armenia Safari 210 Armenia Microsoft Edge 189 Armenia Mozilla Firefox 332 Armenia IOS App 250 Armenia Chrome 304 Armenia Android App 487 Aruba Chrome 228 Aruba IOS App 155 Aruba Microsoft Edge 99 Aruba Android App 234 Aruba Safari 110 Aruba Mozilla Firefox 164 Aruba Others 186 Australia IOS App 300 Australia Safari 240 Australia Android App 575 Australia Mozilla Firefox 400 Australia Chrome 415 Australia Microsoft Edge 216 Australia Others 374 Austria Chrome 418 Austria Mozilla Firefox 408 Austria Android App 576 Austria IOS App 346 Austria Safari 262 Austria Microsoft Edge 243 Austria Others 421 Azerbaijan Chrome 238 Azerbaijan Safari 128 Azerbaijan IOS App 167 Azerbaijan Others 205 Azerbaijan Microsoft Edge 108 Azerbaijan Android App 298 Azerbaijan Mozilla Firefox 200 Bahamas Microsoft Edge 177 Bahamas Android App 432 Bahamas Mozilla Firefox 306 Bahamas Chrome 296 Bahamas IOS App 243 Bahamas Others 292 Bahamas Safari 186 Bahrain Android App 373 Bahrain Mozilla Firefox 270 Bahrain Others 284 Bahrain Microsoft Edge 147 Bahrain Safari 164 Bahrain IOS App 227 Bahrain Chrome 299 Barbados Microsoft Edge 180 Barbados Others 311 Barbados Safari 211 Barbados Android App 482 Barbados Mozilla Firefox 320 Barbados Chrome 343 Barbados IOS App 252 Belarus Android App 410 Belarus IOS App 220 Belarus Microsoft Edge 153 Belarus Mozilla Firefox 275 Belarus Chrome 256 Belarus Others 279 Belarus Safari 170 Belgium Others 394 Belgium Android App 556 Belgium Mozilla Firefox 396 Belgium Safari 250 Belgium IOS App 325 Belgium Microsoft Edge 225 Belgium Chrome 458 Belize Others 376 Belize Android App 523 Belize Microsoft Edge 207 Belize Mozilla Firefox 359 Belize IOS App 294 Belize Chrome 371 Belize Safari 222 Bosnia and Herzegovina Android App 52 Bosnia and Herzegovina Safari 18 Bosnia and Herzegovina Mozilla Firefox 40 Bosnia and Herzegovina Others 15 Bosnia and Herzegovina IOS App 18 Bosnia and Herzegovina Microsoft Edge 9 Bosnia and Herzegovina Chrome 16 Brazil IOS App 331 Brazil Android App 551 Brazil Safari 249 Brazil Mozilla Firefox 385 Brazil Chrome 460 Brazil Microsoft Edge 225 Brazil Others 403 Bulgaria Android App 576 Bulgaria Safari 240 Bulgaria Microsoft Edge 216 Bulgaria Others 382 Bulgaria Mozilla Firefox 399 Bulgaria Chrome 401 Bulgaria IOS App 306 Cabo Verde Mozilla Firefox 12 Cabo Verde Others 18 Cabo Verde IOS App 28 Cabo Verde Chrome 14 Cabo Verde Microsoft Edge 12 Canada Safari 250 Canada Chrome 377 Canada Mozilla Firefox 372 Canada IOS App 296 Canada Microsoft Edge 213 Canada Android App 552 Canada Others 376 Chile Android App 575 Chile Microsoft Edge 216 Chile Chrome 457 Chile Others 396 Chile Mozilla Firefox 400 Chile Safari 240 Chile IOS App 320 Colombia Mozilla Firefox 386 Colombia IOS App 326 Colombia Safari 240 Colombia Microsoft Edge 225 Colombia Others 399 Colombia Android App 537 Colombia Chrome 401 Costa Rica Microsoft Edge 180 Costa Rica Chrome 357 Costa Rica Safari 210 Costa Rica Mozilla Firefox 324 Costa Rica Android App 483 Costa Rica Others 336 Costa Rica IOS App 270 Croatia Mozilla Firefox 246 Croatia Safari 150 Croatia IOS App 198 Croatia Chrome 255 Croatia Microsoft Edge 138 Croatia Android App 345 Croatia Others 240 Cuba IOS App 198 Cuba Chrome 258 Cuba Safari 180 Cuba Mozilla Firefox 255 Cuba Microsoft Edge 156 Cuba Others 270 Cuba Android App 411 Cyprus Mozilla Firefox 171 Cyprus Microsoft Edge 84 Cypr